pkg/resource/plugin/provider_plugin.go

```diff
@@ -257,10 +257,14 @@ func (p *provider) Diff(urn resource.URN, id resource.ID,
 		return DiffResult{}, err
 	}
 
-	// If this function is called, we must have complete configuration for the underlying provider. Per DiffConfig,
-	// any unknown input will cause the provider to be replaced, which will cause all of its resources to be replaced,
-	// and we do not call `Diff` for resources that are being replaced due to a change to their provider reference.
-	contract.Assert(p.cfgknown)
+	// If the configuration for this provider was not fully known--e.g. if we are doing a preview and some input
+	// property was sourced from another resource's output properties--don't call into the underlying provider.
+	// Instead, indicate that the diff is unavailable and write a message
+	if !p.cfgknown {
+		logging.V(7).Infof("%s: cannot diff due to unknown config", label)
+		err := errors.New("cannot diff: this resource's provider has unknown configuration values")
```

Comment:
"cannot diff" sounds like an error and not a warning and this warning isn't actionable for users. Could we do something like:

"This resource provider has inputs that are not known. Diffs of resources using this provider may be incorrect or unreliable."

Reply:
I originally had something like that, but I wanted to get the important bit--i.e. that the diff of the resource may not be accurate--at the front of the message.

Reply:
I don't think `"cannot diff: this resource's provider has unknown configuration values"` accomplishes that, though - that message doesn't tell me that the diff won't be accurate. It just tell me that we couldn't diff it, which is a jargon-y way to say that it might not be accurate. I think it's better to explicitly say "might not be accurate" somewhere in the message.

Reply:
Here are some other candidates:
`diff unavailable: ...`
`diff unknown: ...`
`unreliable diff: ...`
`diff may be unreliable: ...`

Reply:
Yeah - I was about to suggest almost exactly what @swgillespie mentioned above after watching the screencast before seeing this thread.  I don't think "cannot diff" is clear at a level that users will understand.  I feel like this is an implementation details that is a layer or two beneath what the user can reason about.

As a slight play on @swgillespie's suggestion, what about:

"The provider for this resource has inputs that are not known during preview. Diffs of resources using this provider may not correctly capture the changes that will be applied during an update."

Reply:
@pgavlin After seeing the implementation though - I'm not 100% sure that this error can only be reported under the conditions described in my proposed message above.  Is that your concern, that this error can happen in other cases?  If so, what are those other cases?

Reply:
How about "The provider for this resource has inputs that are not known during preview. This preview may not correctly represent the changes that will be applied during an update."

Reply:
> Is that your concern, that this error can happen in other cases?

Nope, the only time this should happen is during a preview. All values should be known at deploy time.